### PR TITLE
Make the validity period of cert configurable

### LINKED ISSUE #2985: The expiration time of the cert
The expiration time of the cert by  kubectl-karmada is 1 year

``` golang
func (i *CommandInitOption) genCerts() error {
	notAfter := time.Now().Add(cert.Duration365d).UTC()
        ......
}
```

The expiration time of the cert by  deploy-karmada.sh is 10 years

```  shell
function util::create_signing_certkey {
    ......
    ${OPENSSL_BIN} req -x509 -sha256 -new -nodes -days 3650 -newkey rsa:2048 -keyout "${dest_dir}/${id}.key" -out "${dest_dir}/${id}.crt" -subj "/CN=${cn}/"
    ......
}
``` 

Expectations are 10 years

### PR DESCRIPTION
**Pull request by lonelyCZ**
Signed-off-by: lonelyCZ <chengzhe@zju.edu.cn>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #2985 

**Special notes for your reviewer**:
This is continuing develoption of #2987 that has been hold too long. And it is useful for users.

What do you think? Solfly ping @hexiaofeng 

**Does this PR introduce a user-facing change?**:

```release-note
karmadactl: import `--cert-validity-period` for `init` to make the validity period of cert configurable
```

### CODE REVIEW

File: pkg/karmadactl/cmdinit/cmdinit.go [hunk 1]
```diff
@@ -102,6 +103,7 @@ func NewCmdInit(parentCommand string) *cobra.Command {
 	// cert
 	flags.StringVar(&opts.ExternalIP, "cert-external-ip", "", "the external IP of Karmada certificate (e.g 192.168.1.2,172.16.1.2)")
 	flags.StringVar(&opts.ExternalDNS, "cert-external-dns", "", "the external DNS of Karmada certificate (e.g localhost,localhost.com)")
+	flags.DurationVar(&opts.ValidityPeriod, "cert-validity-period", cert.Duration365d, "the validity period of Karmada certificate (e.g 8760h0m0s: default 365d)")
```
Comment (RainbowMango) [inline pkg/karmadactl/cmdinit/cmdinit.go:106]:
```suggestion
	flags.DurationVar(&opts.ValidityPeriod, "cert-validity-period", cert.Duration365d, "The validity period of Karmada certificate (e.g 8760h0m0s: default 365d)")
```
Does `Karmada cerificate` means all signed certificates, including both `client` and `server`?

By the way, we don't need to specify the default value again after we set the default value. The framework would set the default value.

Comment (RainbowMango) [reply pkg/karmadactl/cmdinit/cmdinit.go:106]: > e.g 8760h0m0s
Can people set the flag by `year` /`month`/`day`?

Comment (RainbowMango) [reply pkg/karmadactl/cmdinit/cmdinit.go:106]: I suggest renaming the `cert.Duration365d` to `cert.DefaultCertificateValidity`.

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go [hunk 2]
```diff
@@ -121,6 +121,7 @@ type CommandInitOption struct {
 	CRDs                               string
 	ExternalIP                         string
 	ExternalDNS                        string
+	ValidityPeriod                     time.Duration
```
Comment (RainbowMango) [inline pkg/karmadactl/cmdinit/kubernetes/deploy.go:124]:
```suggestion
	CertificateValidity                     time.Duration
```
How about this?

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go [hunk 3]
```diff
@@ -218,7 +219,7 @@ func initializeDirectory(path string) error {
 
 // genCerts create ca etcd karmada cert
 func (i *CommandInitOption) genCerts() error {
-	notAfter := time.Now().Add(cert.Duration365d).UTC()
```
Comment (RainbowMango) [inline pkg/karmadactl/cmdinit/kubernetes/deploy.go:221]: There is more than one place referencing this default variable,  please check if we get it from the option.

File: pkg/karmadactl/cmdinit/cmdinit.go (hunk 1 above)
Comment (lonelyCZ) [reply pkg/karmadactl/cmdinit/cmdinit.go:106]: > Does Karmada cerificate means all signed certificates, including both client and server?

Yes, include certs of all components.

> Can people set the flag by `year` /`month`/`day`?

It seems not working, it is only recognizing the default format(hour/minute/second) of time.Duration. It should be more flexible about fine grained time.

> By the way, we don't need to specify the default value again after we set the default value. The framework would set the default value.

Yes, but `(e.g 8760h0m0s: default 365d)` is only for commenting `8760h0m0s == 365d`.

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go (hunk 3 above)
Comment (lonelyCZ) [reply pkg/karmadactl/cmdinit/kubernetes/deploy.go:221]: Thanks for pointing that, it has some problems and I am going to fix it.

File: pkg/karmadactl/cmdinit/cmdinit.go (hunk 1 above)
Comment (RainbowMango) [reply pkg/karmadactl/cmdinit/cmdinit.go:106]: OK. get it. Please echo the output of `karmadactl init --help`, especially for this flag part.
Maybe we can comment it like `e.g 8760h0m0s, that is 365d`.

Comment (lonelyCZ) [reply pkg/karmadactl/cmdinit/cmdinit.go]:
```
Options:
    --cert-external-dns='':
        the external DNS of Karmada certificate (e.g localhost,localhost.com)

    --cert-external-ip='':
        the external IP of Karmada certificate (e.g 192.168.1.2,172.16.1.2)

    --cert-validity-period=8760h0m0s:
        the validity period of Karmada certificate (e.g 8760h0m0s, that is 365 days)
```

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go (hunk 3 above)
Comment (lonelyCZ) [reply pkg/karmadactl/cmdinit/kubernetes/deploy.go:221]: `Duration365d` is used here, and it only work when `cfg.NotAfter == nil`, but the `cfg.NotAfter` has been filled when `NewCertConfig`.

https://github.com/karmada-io/karmada/blob/b1f8b09d073d9f856aef90297e5ae3f7c780ee00/pkg/karmadactl/cmdinit/cert/cert.go#L110-L113

https://github.com/karmada-io/karmada/blob/b1f8b09d073d9f856aef90297e5ae3f7c780ee00/pkg/karmadactl/cmdinit/cert/cert.go#L234-L244

https://github.com/karmada-io/karmada/blob/b1f8b09d073d9f856aef90297e5ae3f7c780ee00/pkg/karmadactl/cmdinit/kubernetes/deploy.go#L233-L234

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**